### DISCUSSION TITLE
sqitch rework: revert to previous version (same hash) violates constraint

### DISCUSSION THREAD

**Discussion by rossocampbell** [theory/sqitch#750] [General]
After doing a sqitch rework and modifying a routine to match its previous version, on deploy I get 

ERROR:  duplicate key value violates unique constraint "changes_project_script_hash_key"

Is there a way to revert to a previous version of code or am I stuck with having to change some white space or a comment?

**Comment by theory** [comment]
If you change it to be identical to the previous version, nothing is being reworked. The way to do it is to delete the reworked change. Unless you have some reason to keep it (e.g., already released?). In which case either rework it again to change it back, or put a wayward space in the reworked deploy script so that the SHA hash ID turns out different.

**Comment by rossocampbell** [comment]
I created a small test case to demonstrate, [sqitch-rework-test.zip](https://github.com/sqitchers/sqitch/files/11464659/sqitch-rework-test.zip).


It consists of a single change, func1. So I did the following:

1. sqitch add func1
2. sqitch tag tag1
3. sqitch rework func1 (simulating modification of func1 that gets released)
4. sqitch tag tag2
5. sqitch rework func1 (simulating "didn't like that change, revert it to the original")

Now it gets the error (see below). Is this intended behavior? If so, I'll do the white space thing and get on with my life :-).


```$ sqitch deploy
Deploying changes to testdb
  + func1 @tag1 .. ok
  + func1 @tag2 .. ok
  + func1 ........ ERROR:  duplicate key value violates unique constraint "changes_project_script_hash_key"
DETAIL:  Key (project, script_hash)=(testdb, a015ae1b77b102211919c8c50a6b2a2572fa275d) already exists.
not ok
Reverting all changes
  - func1 @tag2 .. ok
  - func1 @tag1 .. ok
Deploy failed```

**Comment by theory** [comment]
it is intended behavior. Stick an extra space or new line in the `"didn't like that change, revert it to the original"` version so that it will not be byte-for-byte identical to the previous version of the deploy script.

**Comment by rossocampbell** [comment]
Great, thanks!
